### PR TITLE
Adds a role based authentication backend for the experimental API

### PR DESCRIPTION
**Pull request by jhtimmins**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/api/auth/backend/role_based_auth.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+T = TypeVar("T", bound=Callable)  # pylint: disable=invalid-name
+
+
+def requires_authentication(function):
+    """Decorator for functions that require authentication"""
+
+    @wraps(function)
+    def decorated(*args, **kwargs):
+        appbuilder = cached_app().appbuilder  # pylint: disable=no-member
```
Comment (mik-laj) [inline airflow/api/auth/backend/role_based_auth.py]:
```suggestion
        appbuilder = current_app.appbuilder  # pylint: disable=no-member
```
Please use Flask.current_app. This makes it harder to create side effects.

**Comment by mik-laj** [comment]
What do you think to add role support for all auth backend? I don't understand why you want to do it on the authentication level and not as a separate decorator that will be responsible for permissions handling.

**Comment by mik-laj** [comment]
Can you add some documentation? https://airflow.readthedocs.io/en/latest/security.html#api-authentication

File: tests/www/test_role_based_auth.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                delete_dag.delete_dag(dag)
+            except DagNotFound:
+                pass
+
+    def create_test_dag(self, dag_id):
+        dag = DagModel(dag_id=dag_id)
+        self.session.add(dag)
+        self.session.commit()
+
+    def use_admin_role(self):
```
Comment (mik-laj) [inline tests/www/test_role_based_auth.py:112]: How is the end user to set this attribute?

Comment (jhtimmins) [reply tests/www/test_role_based_auth.py:112]: This isn't accessible to users. It just allows the tests to toggle between Admin access and Viewer access.

**Comment by jhtimmins** [comment]
> What do you think to add role support for all auth backend? I don't understand why you want to do it on the authentication level and not as a separate decorator that will be responsible for permissions handling.

We did it this way for the sake of simplicity. Rather than adding an additional hook for an API that's going to be deprecated shortly, this lets us hook into existing functionality in a non-disruptive way.

**Comment by mik-laj** [comment]
How can I configure your auth backend and authorization with Google OpenID at the same time?

**Review by mik-laj — CHANGES_REQUESTED** [review]
I don't understand how this is supposed to work.

**Comment by jhtimmins** [comment]
@mik-laj Your concerns make sense. I'm going to close this ticket and will move it into an internal repo of ours. Thanks for reviewing this!

**Pull request closed without merging**